sdk/go/common/workspace/plugins.go

```diff
@@ -344,6 +348,98 @@ func (source *gitlabSource) URL() string {
 	return fmt.Sprintf("gitlab://%s/%s", source.host, source.project)
 }
 
+type gitSource struct {
+	url  string
+	path string
+
+	cloneAndCheckout func(context.Context, string, plumbing.Hash, string) error
+	cloneOrPull      func(context.Context, string, plumbing.ReferenceName, string, bool) error
+}
+
+func newGitHTTPSSource(url *url.URL) (*gitSource, error) {
+	url.Scheme = "https"
+	u, path, err := gitutil.ParseGitRepoURL(url.String())
+	if err != nil {
+		return nil, err
+	}
+	if path != "" {
+		return nil,
+			// TODO[pulumi/pulumi#18250]: We currently don't support installing plugins from subdirectories.
+			//nolint:lll
+			errors.New("cannot install git plugin from subdirectory.  See https://github.com/pulumi/pulumi/issues/18250")
+	}
+	return &gitSource{
+		url:              u,
+		path:             path,
+		cloneAndCheckout: gitutil.GitCloneAndCheckoutCommit,
+		cloneOrPull:      gitutil.GitCloneOrPull,
+	}, nil
+}
+
+func (source *gitSource) GetLatestVersion(
+	ctx context.Context, _ func(*http.Request) (io.ReadCloser, int64, error),
+) (*semver.Version, error) {
+	// TODO: support this for the users convenience
+	return &semver.Version{},
+		errors.New("GetLatestVersion is not supported for git sources, please specify the version of the plugin")
+}
+
+func (source *gitSource) Download(
```

Comment:
Maybe just me, but it might be good to have a doc comment for this that just outlines the general flow of how `Download` works for Git sources. E.g.

* It's a version of the form 0.0.0-x<hash> we do X
* If it's 0.0.0 with no pre-release it's HEAD
* etc.

I feel like it'd give the reader a way to quickly navigate the `if`-chain a bit better, and also capture any assumptions, like where we prepend `v` to versions etc.